cpp/src/arrow/ipc/writer.cc

```diff
@@ -329,15 +329,24 @@ class RecordBatchSerializer {
       return Status::OK();
     }
 
-    int64_t required_bytes = sizeof(offset_type) * (array.length() + 1);
-    if (array.value_offset(0) > 0) {
+    const int64_t required_bytes = sizeof(offset_type) * (array.length() + 1);
+
+    offset_type first_offset = 0;
+    RETURN_NOT_OK(MemoryManager::CopyBufferSliceToCPU(
```

Comment:
I think I'll need some context here. Is this function supposed to be able to deal with non-CPU arrays? Does this rule apply to other functions in this source file too?

Reply:
Yes, it's supposed to.